### PR TITLE
sys_lwmutex/mutex: Fix race on lock timeout

### PR DESCRIPTION
**Pull request by elad335**
Fix a race condition causing possible discard of waiter PPU which just registered while another timed-out.

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_mutex.cpp [hunk 1]
```diff
@@ -245,12 +245,38 @@ error_code sys_mutex_lock(ppu_thread& ppu, u32 mutex_id, u64 timeout)
 
 				std::lock_guard lock(mutex->mutex);
 
-				if (!mutex->unqueue(mutex->control.raw().sq, &ppu))
+				bool success = false;
+
+				mutex->control.fetch_op([&](lv2_mutex::control_data_t& data)
 				{
-					break;
+					success = false;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_mutex.cpp:252]: obsolete?

File: rpcs3/Emu/Cell/lv2/sys_lwmutex.cpp [hunk 2]
```diff
@@ -236,12 +236,38 @@ error_code _sys_lwmutex_lock(ppu_thread& ppu, u32 lwmutex_id, u64 timeout)
 
 				std::lock_guard lock(mutex->mutex);
 
-				if (!mutex->unqueue(mutex->lv2_control.raw().sq, &ppu))
+				bool success = false;
+
+				mutex->lv2_control.fetch_op([&](lv2_lwmutex::control_data_t& data)
+				{
+					success = false;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_lwmutex.cpp:243]: obsolete?

File: rpcs3/Emu/Cell/lv2/sys_mutex.cpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_mutex.cpp:252]: Wdym? fetch_op is a loop, it can execute more than once.

**Comment by brian218** [comment]
Could you please confirm if `sys_cond_wait` also needs this fix?

**Comment by elad335** [comment]
No because in sys_cond_wait the timeout only applies on condional waiting. It always trying to the mutex infinitely.

**Comment by Nekotekina** [comment]
Maybe we should put all syscalls under a single mutex as it was multiple years ago. It would simplify the code dramatically and possibly prevent issues like this.

**Comment by elad335** [comment]
Superseded be #13917

**Pull request closed without merging**